File: samples/FsBulletML.Sample.Unity2D.CSharp/Assets/Scripts/UrpPlayModeCompat.cs
using UnityEngine;
using UnityEngine.Rendering;
using UnityEngine.Rendering.Universal;

/// <summary>
/// Runtime safety net so GameObject sprites (player/enemy/bg/bomb/HUD) and the
/// Main Camera actually draw under URP Forward+/Entities Graphics.
/// Persistent asset conversion is done by FixBlackScreen in the Editor.
/// </summary>
public static class UrpPlayModeCompat
{
    static bool _applied;

    public static void Apply()
    {
        if (_applied)
        {
            return;
        }

        _applied = true;
        EnsureCameras();
        EnsureDirectionalLight();
        UpgradeSceneRenderers();
        BindSpriteTextures();
    }

    public static Shader FindUrpUnlit()
    {
        return Shader.Find("Universal Render Pipeline/Unlit")
            ?? Shader.Find("Universal Render Pipeline/Unlit/Unlit");
    }

    public static Shader FindUrpParticleUnlit()
    {
        return Shader.Find("Universal Render Pipeline/Particles/Unlit")
            ?? FindUrpUnlit();
    }

    public static void ConfigureUrpUnlit(Material mat, Texture texture, Color color, bool transparent, bool additive)
    {
        if (mat == null)
        {
            return;
        }

        if (mat.HasProperty("_BaseMap"))
        {
            mat.SetTexture("_BaseMap", texture);
        }

        if (mat.HasProperty("_MainTex"))
        {
            mat.SetTexture("_MainTex", texture);
        }

        mat.mainTexture = texture;

        if (mat.HasProperty("_BaseColor"))
        {
            mat.SetColor("_BaseColor", color);
        }

        if (mat.HasProperty("_Color"))
        {
            mat.SetColor("_Color", color);
        }

        if (mat.HasProperty("_Cull"))
        {
            mat.SetFloat("_Cull", 0f);
        }

        mat.doubleSidedGI = true;

        if (transparent)
        {
            if (mat.HasProperty("_Surface"))
            {
                mat.SetFloat("_Surface", 1f);
            }

            var dst = additive ? BlendMode.One : BlendMode.OneMinusSrcAlpha;
            if (mat.HasProperty("_Blend"))
            {
                mat.SetFloat("_Blend", additive ? 2f : 0f);
            }

            if (mat.HasProperty("_SrcBlend"))
            {
                mat.SetFloat("_SrcBlend", (float)BlendMode.SrcAlpha);
            }

            if (mat.HasProperty("_DstBlend"))
            {
                mat.SetFloat("_DstBlend", (float)dst);
            }

            if (mat.HasProperty("_ZWrite"))
            {
                mat.SetFloat("_ZWrite", 0f);
            }

            mat.SetOverrideTag("RenderType", "Transparent");
            mat.EnableKeyword("_SURFACE_TYPE_TRANSPARENT");
            mat.DisableKeyword("_ALPHAPREMULTIPLY_ON");
            if (additive)
            {
                mat.EnableKeyword("_ALPHAMODULATE_ON");
            }

            mat.renderQueue = (int)RenderQueue.Transparent;
        }
        else
        {
            if (mat.HasProperty("_Surface"))
            {
                mat.SetFloat("_Surface", 0f);
            }

            if (mat.HasProperty("_ZWrite"))
            {
                mat.SetFloat("_ZWrite", 1f);
            }

            mat.SetOverrideTag("RenderType", "Opaque");
            mat.DisableKeyword("_SURFACE_TYPE_TRANSPARENT");
            mat.renderQueue = (int)RenderQueue.Geometry;
        }

        mat.enableInstancing = true;
    }

    public static bool IsUrpShader(Shader shader)
    {
        if (shader == null)
        {
            return false;
        }

        var n = shader.name ?? string.Empty;
        return n.IndexOf("Universal Render Pipeline", System.StringComparison.OrdinalIgnoreCase) >= 0
            || n.IndexOf("Shader Graphs", System.StringComparison.OrdinalIgnoreCase) >= 0;
    }

    static void EnsureCameras()
    {
        Camera main = Camera.main;
        if (main == null)
        {
            var cameras = Object.FindObjectsByType<Camera>(FindObjectsInactive.Include);
            for (int i = 0; i < cameras.Length; i++)
            {
                if (cameras[i] != null && cameras[i].CompareTag("MainCamera"))
                {
                    main = cameras[i];
                    break;
                }
            }

            if (main == null && cameras.Length > 0)
            {
                main = cameras[0];
            }
        }

        if (main == null)
        {
            Debug.LogError("UrpPlayModeCompat: no camera found.");
            return;
        }

        main.enabled = true;
        main.gameObject.tag = "MainCamera";
        main.gameObject.SetActive(true);
        main.clearFlags = CameraClearFlags.SolidColor;
        // Original sample clear was already black; keep it. Scrolling bg mesh fills the view.
        if (main.backgroundColor.a < 0.01f)
        {
            var bg = main.backgroundColor;
            bg.a = 1f;
            main.backgroundColor = bg;
        }

        var mainData = main.GetUniversalAdditionalCameraData();
        mainData.renderType = CameraRenderType.Base;
        mainData.SetRenderer(0);
        mainData.renderPostProcessing = false;
        mainData.cameraStack.Clear();

        var camerasAll = Object.FindObjectsByType<Camera>(FindObjectsInactive.Include);
        for (int i = 0; i < camerasAll.Length; i++)
        {
            var cam = camerasAll[i];
            if (cam == null || cam == main)
            {
                continue;
            }

            cam.enabled = true;
            var data = cam.GetUniversalAdditionalCameraData();
            data.renderType = CameraRenderType.Overlay;
            data.SetRenderer(0);
            data.renderPostProcessing = false;
            if (!mainData.cameraStack.Contains(cam))
            {
                mainData.cameraStack.Add(cam);
            }
        }

        Debug.Log(
            "UrpPlayModeCompat cameras: main=" + main.name
            + " ortho=" + main.orthographic
            + " fov=" + main.fieldOfView
            + " size=" + main.orthographicSize
            + " pos=" + main.transform.position
            + " near=" + main.nearClipPlane
            + " far=" + main.farClipPlane
            + " mask=" + main.cullingMask
            + " stack=" + mainData.cameraStack.Count
            + " pipeline=" + (GraphicsSettings.defaultRenderPipeline != null ? GraphicsSettings.defaultRenderPipeline.name : "NULL"));
    }

    static void EnsureDirectionalLight()
    {
        var lights = Object.FindObjectsByType<Light>(FindObjectsInactive.Include);
        for (int i = 0; i < lights.Length; i++)
        {
            if (lights[i] != null && lights[i].type == LightType.Directional)
            {
                return;
            }
        }

        var go = new GameObject("Directional Light");
        var light = go.AddComponent<Light>();
        light.type = LightType.Directional;
        light.intensity = 1f;
        light.color = Color.white;
        go.transform.rotation = Quaternion.Euler(50f, -30f, 0f);
    }

    static void UpgradeSceneRenderers()
    {
        var unlit = FindUrpUnlit();
        var particle = FindUrpParticleUnlit();
        if (unlit == null)
        {
            Debug.LogError("UrpPlayModeCompat: URP Unlit shader not found.");
            return;
        }

        var renderers = Object.FindObjectsByType<Renderer>(FindObjectsInactive.Include);
        for (int i = 0; i < renderers.Length; i++)
        {
            var r = renderers[i];
            if (r == null)
            {
                continue;
            }

            var shared = r.sharedMaterials;
            if (shared == null)
            {
                continue;
            }

            Material[] instances = null;
            for (int m = 0; m < shared.Length; m++)
            {
                var src = shared[m];
                if (src == null || IsUrpShader(src.shader))
                {
                    continue;
                }

                if (instances == null)
                {
                    instances = r.materials;
                }

                var shaderName = src.shader != null ? src.shader.name : string.Empty;
                bool additive = shaderName.IndexOf("Particle", System.StringComparison.OrdinalIgnoreCase) >= 0
                    || shaderName.IndexOf("Additive", System.StringComparison.OrdinalIgnoreCase) >= 0;
                var shader = additive && particle != null ? particle : unlit;
                var copy = new Material(shader) { name = src.name + " (URP)" };
                var tex = src.HasProperty("_MainTex") ? src.GetTexture("_MainTex") : src.mainTexture;
                var color = src.HasProperty("_Color") ? src.GetColor("_Color") : Color.white;
                bool transparent = !(r is MeshRenderer && r.gameObject.name.IndexOf("background", System.StringComparison.OrdinalIgnoreCase) >= 0);
                ConfigureUrpUnlit(copy, tex != null ? tex : Texture2D.whiteTexture, color, transparent, additive);
                instances[m] = copy;
            }

            if (instances != null)
            {
                r.materials = instances;
            }
        }
    }

    static void BindSpriteTextures()
    {
        var sprites = Object.FindObjectsByType<SpriteRenderer>(FindObjectsInactive.Include);
        for (int i = 0; i < sprites.Length; i++)
        {
            BindSprite(sprites[i]);
        }
    }

    public static void BindSprite(SpriteRenderer sr)
    {
        if (sr == null || sr.sprite == null)
        {
            return;
        }

        var tex = sr.sprite.texture;
        if (tex == null)
        {
            return;
        }

        var block = new MaterialPropertyBlock();
        sr.GetPropertyBlock(block);
        block.SetTexture("_MainTex", tex);
        block.SetTexture("_BaseMap", tex);
        block.SetColor("_BaseColor", sr.color);
        block.SetColor("_Color", sr.color);
        sr.SetPropertyBlock(block);
    }
}

File: samples/FsBulletML2.Sample.Unity2D.FSharp/FsBulletML2.Sample.Unity2D.FSharp/UrpPlayModeCompat.fs
namespace FsBulletML2.Sample.Unity2D.FSharp

open System
open UnityEngine
open UnityEngine.Rendering
open UnityEngine.Rendering.Universal

/// URP に移したあと、GameObject の見た目が出るようにする安全網。
[<AbstractClass; Sealed>]
type UrpPlayModeCompat private () =

  static let mutable applied = false

  /// URP の Unlit。名前が版で変わるので 2 つ 試す
  static member FindUrpUnlit () =
    match Shader.Find "Universal Render Pipeline/Unlit" with
    | null -> Shader.Find "Universal Render Pipeline/Unlit/Unlit"
    | s -> s

  static member FindUrpParticleUnlit () =
    match Shader.Find "Universal Render Pipeline/Particles/Unlit" with
    | null -> UrpPlayModeCompat.FindUrpUnlit ()
    | s -> s

  /// そのシェーダが既に URP のものか。差し替えを 2 度 やらないため
  static member IsUrpShader (shader: Shader) =
    if isNull (box shader) then false
    else
      let n = if isNull shader.name then "" else shader.name
      n.IndexOf("Universal Render Pipeline", StringComparison.OrdinalIgnoreCase) >= 0
      || n.IndexOf("Shader Graphs", StringComparison.OrdinalIgnoreCase) >= 0

  /// URP の Unlit として描けるように材質を整える
  static member ConfigureUrpUnlit (mat: Material, texture: Texture, color: Color,
                                   transparent: bool, additive: bool) =
    if isNull (box mat) then () else

    if mat.HasProperty "_BaseMap" then mat.SetTexture("_BaseMap", texture)
    if mat.HasProperty "_MainTex" then mat.SetTexture("_MainTex", texture)
    mat.mainTexture <- texture
    if mat.HasProperty "_BaseColor" then mat.SetColor("_BaseColor", color)
    if mat.HasProperty "_Color" then mat.SetColor("_Color", color)
    if mat.HasProperty "_Cull" then mat.SetFloat("_Cull", 0.0f)
    mat.doubleSidedGI <- true

    if transparent then
      if mat.HasProperty "_Surface" then mat.SetFloat("_Surface", 1.0f)
      let dst = if additive then BlendMode.One else BlendMode.OneMinusSrcAlpha
      if mat.HasProperty "_Blend" then mat.SetFloat("_Blend", (if additive then 2.0f else 0.0f))
      if mat.HasProperty "_SrcBlend" then mat.SetFloat("_SrcBlend", float32 BlendMode.SrcAlpha)
      if mat.HasProperty "_DstBlend" then mat.SetFloat("_DstBlend", float32 dst)
      if mat.HasProperty "_ZWrite" then mat.SetFloat("_ZWrite", 0.0f)
      mat.SetOverrideTag("RenderType", "Transparent")
      mat.EnableKeyword "_SURFACE_TYPE_TRANSPARENT"
      mat.DisableKeyword "_ALPHAPREMULTIPLY_ON"
      if additive then mat.EnableKeyword "_ALPHAMODULATE_ON"
      mat.renderQueue <- int RenderQueue.Transparent
    else
      if mat.HasProperty "_Surface" then mat.SetFloat("_Surface", 0.0f)
      if mat.HasProperty "_ZWrite" then mat.SetFloat("_ZWrite", 1.0f)
      mat.SetOverrideTag("RenderType", "Opaque")
      mat.DisableKeyword "_SURFACE_TYPE_TRANSPARENT"
      mat.renderQueue <- int RenderQueue.Geometry

    mat.enableInstancing <- true

  /// カメラを URP が描ける形にする。
  static member private EnsureCameras () =
    let cameras = UnityEngine.Object.FindObjectsByType<Camera>(FindObjectsInactive.Include)
    let main =
      match Camera.main with
      | null ->
        let tagged = cameras |> Array.tryFind (fun c -> not (isNull (box c)) && c.CompareTag "MainCamera")
        match tagged with
        | Some c -> c
        | None -> if cameras.Length > 0 then cameras.[0] else null
      | c -> c

    if isNull (box main) then
      Debug.LogError "UrpPlayModeCompat: カメラが 1 台 も無い"
    else
      main.enabled <- true
      main.gameObject.tag <- "MainCamera"
      main.gameObject.SetActive true
      main.clearFlags <- CameraClearFlags.SolidColor
      // 元のサンプルの背景は黒。透明のままだと何も塗られないので不透明にする
      if main.backgroundColor.a < 0.01f then
        let mutable bg = main.backgroundColor
        bg.a <- 1.0f
        main.backgroundColor <- bg

      let mainData = main.GetUniversalAdditionalCameraData()
      mainData.renderType <- CameraRenderType.Base
      mainData.SetRenderer 0
      mainData.renderPostProcessing <- false
      mainData.cameraStack.Clear()

      for cam in cameras do
        if not (isNull (box cam)) && not (Object.ReferenceEquals(cam, main)) then
          cam.enabled <- true
          let data = cam.GetUniversalAdditionalCameraData()
          data.renderType <- CameraRenderType.Overlay
          data.SetRenderer 0
          data.renderPostProcessing <- false
          if not (mainData.cameraStack.Contains cam) then
            mainData.cameraStack.Add cam

      Debug.Log(
        sprintf "UrpPlayModeCompat cameras: main=%s ortho=%b size=%f stack=%d pipeline=%s"
          main.name main.orthographic main.orthographicSize mainData.cameraStack.Count
          (if isNull (box GraphicsSettings.defaultRenderPipeline) then "NULL"
           else GraphicsSettings.defaultRenderPipeline.name))

  /// 平行光源が無ければ足す。URP の Lit は光源が無いと真っ黒になる
  static member private EnsureDirectionalLight () =
    let lights = UnityEngine.Object.FindObjectsByType<Light>(FindObjectsInactive.Include)
    // `type` は F# の予約語なので、メンバ名はバッククォートで囲む
    let hasDirectional =
      lights |> Array.exists (fun l -> not (isNull (box l)) && l.``type`` = LightType.Directional)
    if not hasDirectional then
      let go = new GameObject("Directional Light")
      let light = go.AddComponent<Light>()
      light.``type`` <- LightType.Directional
      light.intensity <- 1.0f
      light.color <- Color.white
      go.transform.rotation <- Quaternion.Euler(50.0f, -30.0f, 0.0f)

  /// Built-in のシェーダを使っている材質を URP の Unlit へ差し替える。
  static member private UpgradeSceneRenderers () =
    let unlit = UrpPlayModeCompat.FindUrpUnlit ()
    let particle = UrpPlayModeCompat.FindUrpParticleUnlit ()
    if isNull (box unlit) then
      Debug.LogError "UrpPlayModeCompat: URP の Unlit シェーダが見つからない"
    else
      let renderers = UnityEngine.Object.FindObjectsByType<Renderer>(FindObjectsInactive.Include)
      for r in renderers do
        if not (isNull (box r)) then
          let shared = r.sharedMaterials
          if not (isNull shared) then
            let mutable instances : Material[] = null
            for m in 0 .. shared.Length - 1 do
              let src = shared.[m]
              if not (isNull (box src)) && not (UrpPlayModeCompat.IsUrpShader src.shader) then
                if isNull instances then instances <- r.materials
                let shaderName = if isNull (box src.shader) then "" else src.shader.name
                let additive =
                  shaderName.IndexOf("Particle", StringComparison.OrdinalIgnoreCase) >= 0
                  || shaderName.IndexOf("Additive", StringComparison.OrdinalIgnoreCase) >= 0
                let shader = if additive && not (isNull (box particle)) then particle else unlit
                let copy = new Material(shader, name = src.name + " (URP)")
                let tex = if src.HasProperty "_MainTex" then src.GetTexture "_MainTex" else src.mainTexture
                let color = if src.HasProperty "_Color" then src.GetColor "_Color" else Color.white
                // 背景の板だけ不透明。ほかは透過で重ねる
                let transparent =
                  not (r :? MeshRenderer
                       && r.gameObject.name.IndexOf("background", StringComparison.OrdinalIgnoreCase) >= 0)
                let tex = if isNull (box tex) then Texture2D.whiteTexture :> Texture else tex
                UrpPlayModeCompat.ConfigureUrpUnlit(copy, tex, color, transparent, additive)
                instances.[m] <- copy
            if not (isNull instances) then r.materials <- instances

  /// SpriteRenderer のテクスチャを property block で束ね直す。
  /// URP の sprite 経路はこれが無いと白（またはピンク）になる
  static member BindSprite (sr: SpriteRenderer) =
    if not (isNull (box sr)) && not (isNull (box sr.sprite)) then
      let tex = sr.sprite.texture
      if not (isNull (box tex)) then
        let block = new MaterialPropertyBlock()
        sr.GetPropertyBlock block
        block.SetTexture("_MainTex", tex)
        block.SetTexture("_BaseMap", tex)
        block.SetColor("_BaseColor", sr.color)
        block.SetColor("_Color", sr.color)
        sr.SetPropertyBlock block

  static member private BindSpriteTextures () =
    let sprites = UnityEngine.Object.FindObjectsByType<SpriteRenderer>(FindObjectsInactive.Include)
    for sr in sprites do UrpPlayModeCompat.BindSprite sr

  /// Play の頭で 1 回 だけ。 2 度 呼んでも 2 度目 は何もしない
  static member Apply () =
    if not applied then
      applied <- true
      UrpPlayModeCompat.EnsureCameras ()
      UrpPlayModeCompat.EnsureDirectionalLight ()
      UrpPlayModeCompat.UpgradeSceneRenderers ()
      UrpPlayModeCompat.BindSpriteTextures ()
